lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,311 @@
+                lambda x:
+                self._get_task_activation_mean_capacities(
+                    task, x['start'], x['end']),
+                axis=1, result_type='expand')
+
+        phases = [(0, 0, 0)] + self._get_phases(self.plat_info)
+
+        df[['uclamp_val', 'util', 'band']] = df.apply(
+                lambda x: [phases[int(x.phase)][0],
+                           phases[int(x.phase)][1],
+                           phases[int(x.phase)][2]],
+                axis=1, result_type='expand')
+
+        return df
+
+    @classmethod
+    def _test_passed(cls, df, col, minmax_func):
+        minmax_df = df.apply(minmax_func, axis=1, result_type='expand')
+        passed = df[col].between(minmax_df[0], minmax_df[1])
+
+        return ResultBundle.from_bool(passed.all())
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_placement(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement is compatible with
+        UtilClamp requirements. This is done by comparing the maximum capacity
+        of the CPU on which the task has been placed, with the UtilClamp
+        value.
+        """
+
+        df = self._get_test_df()
+        capacity_margin = self.CAPACITY_MARGIN
+        max_capacities = sorted(set(
+            self.plat_info['cpu-capacities']['rtapp'].values()))
+
+        def expected_minmax(row):
+            uclamp_val = row['uclamp_val']
+            expected = 1024
+
+            for cap in max_capacities:
+                if (cap * capacity_margin) > uclamp_val:
+                    expected = cap
+                    break
+
+            return (expected, expected)
+
+        return self._test_passed(df, 'capacity_max', expected_minmax)
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_freq_selection(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement and frequency selection
+        is compatible with UtilClamp requirements. This is done by comparing
+        the current CPU capacity on which the task has been placed, with the
+        UtilClamp value.
+
+        The expected capacity is :
+          - min: The first capacity below the UtilClamp value.
+          - max: The estimated selected schedutil capacity.
+        """
+
+        df = self._get_test_df()
+        capacities = self._collect_capacities_flatten(self.plat_info)
+        cpu_capacities = self._collect_capacities(self.plat_info)
+
+        def schedutil_map_util_cap(cpu, util):
+            """
+            Returns, for a given util on a given CPU, the capacity that
+            schedutil would select.
+            """
+            max_freq = max(cpu_capacities[cpu])
+            max_cap = max(cpu_capacities[cpu].values())
+            schedutil_freq = (max_freq + (max_freq >> 2)) * util / max_cap
```

Comment:
We can use readable code here (and tbf, the kernel can as well: https://godbolt.org/z/38v8aW ...):
```
			 schedutil_freq = max_freq * (1/self.CAPACITY_MARGIN) * util / max_cap
```